### PR TITLE
raftstore: gc uninitialized stale peer after merge

### LINKED ISSUE #15919: Stale peers are left forever after region merge
## Bug Report

A "stale peer" refers to a peer that still exists on a TiKV node but has been
removed from the raft group, typically through a confchange operation.
TiKV performs regular checks and validations on its peers to ensure that no such
stale peer exists.

However, the presence of the TODO below creates a situation where stale peers
cannot be removed after the corresponding region is merged.

https://github.com/tikv/tikv/blob/e0fe14d57136f645457bcf14ee5ae1a478be04b8/components/raftstore/src/store/worker/pd.rs#L1569-L1572

**It wastes cpu and memory and blocks resolved ts.**

<!--

pd
```log
[2023/11/01 17:41:41.379 +08:00] [INFO] [region.go:735] ["region Version changed"] [region-id=906] [detail="StartKey Changed:{7480000000000000FF775F72800000001EFF143B440000000000FA} -> {7480000000000000FF775F72800000001DFFCFC94A0000000000FA}, EndKey:{7480000000000000FF7800000000000000F8}"] [old-version=171] [new-version=175]
[2023/11/01 17:41:41.600 +08:00] [INFO] [operator_controller.go:464] ["add operator"] [region-id=906] [operator="\"merge-region {merge: region 906 to 1747} (kind:merge,region,leader, region:906(175, 31), createAt:2023-11-01 17:41:41.60077071 +0800 CST m=+11383.180666476, startAt:0001-01-01 00:00:00 +0000 UTC, currentStep:0, size:2, steps:[0:{add learner peer 43228 on store 1}, 1:{add learner peer 43227 on store 343}, 2:{transfer leader from store 313 to store 13}, 3:{use joint consensus, promote learner peer 43228 on store 1 to voter, promote learner peer 43227 on store 343 to voter, demote voter peer 23838 on store 12 to learner, demote voter peer 909 on store 313 to learner}, 4:{leave joint state, promote learner peer 43228 on store 1 to voter, promote learner peer 43227 on store 343 to voter, demote voter peer 23838 on store 12 to learner, demote voter peer 909 on store 313 to learner}, 5:{remove peer on store 12}, 6:{remove peer on store 313}, 7:{merge region 906 into region 1747}], timeout:[43m0s])\""] [additional-info="{\"related merge region\":\"1747\"}"]
[2023/11/01 17:41:42.064 +08:00] [INFO] [region.go:735] ["region Version changed"] [region-id=1747] [detail="StartKey Changed:{7480000000000000FF7800000000000000F8} -> {7480000000000000FF775F72800000001DFFCFC94A0000000000FA}, EndKey:{7480000000000000FF785F728000000000FF0017240000000000FA}"] [old-version=96] [new-version=177]
```

tikv 7
```
[2023/11/01 15:45:34.900 +08:00] [INFO] [peer.rs:321] ["replicate peer"] [store_id=304] [peer_id=1537] [region_id=906] [thread_id=0x5]
[2023/11/01 15:45:34.900 +08:00] [INFO] [raft.rs:2660] ["switched to configuration"] [config="Configuration { voters: Configuration { incoming: Configuration { voters: {} }, outgoing: Configuration>[2023/11/01 15:45:34.900 +08:00] [INFO] [raft.rs:1127] ["became follower at term 0"] [term=0] [raft_id=1537] [region_id=906] [thread_id=0x5]
[2023/11/01 15:45:34.900 +08:00] [INFO] [raft.rs:388] [newRaft] [peers="Configuration { incoming: Configuration { voters: {} }, outgoing: Configuration { voters: {} } }"] ["last term"=0] ["last ind>[2023/11/01 15:45:34.900 +08:00] [INFO] [raw_node.rs:315] ["RawNode created with id 1537."] [id=1537] [raft_id=1537] [region_id=906] [thread_id=0x5]
[2023/11/01 15:45:34.900 +08:00] [INFO] [raft.rs:1362] ["received a message with higher term from 909"] ["msg type"=MsgHeartbeat] [message_term=7] [term=0] [from=909] [raft_id=1537] [region_id=906]>[2023/11/01 15:45:34.900 +08:00] [INFO] [raft.rs:1127] ["became follower at term 7"] [term=7] [raft_id=1537] [region_id=906] [thread_id=0x5]
[2023/11/01 16:03:47.299 +08:00] [INFO] [endpoint.rs:622] ["the max gap of follower resolved-ts is large"] [oldest_candidate="Some(ReadState { idx: 6532, ts: 445336706315190279 })"] [latest_candida>[2023/11/01 16:03:57.300 +08:00] [INFO] [endpoint.rs:622] ["the max gap of follower resolved-ts is large"] [oldest_candidate="Some(ReadState { idx: 6532, ts: 445336706315190279 })"] [latest_candida>

[2023/11/01 18:05:34.997 +08:00] [WARN] [peer.rs:6382] ["leader missing longer than max_leader_missing_duration. To check with pd and other peers whether it's still valid"] [expect=2h] [peer_id=153>
```

tikv 6
```
[2023/11/01 16:00:23.902 +08:00] [WARN] [peer_storage.rs:515] ["failed to try generating snapshot"] [request_peer=1537] [times=1] [peer_id=909] [region_id=906] [thread_id=0x5]
[2023/11/01 16:00:23.902 +08:00] [INFO] [peer_storage.rs:547] ["requesting snapshot"] [request_peer=1537] [request_index=0] [peer_id=909] [region_id=906] [thread_id=0x5]
[2023/11/01 16:01:15.880 +08:00] [INFO] [peer.rs:1307] ["deleting compacted snap file"] [snap_file=906_7_73526] [peer_id=909] [region_id=906] [thread_id=0x5]
[2023/11/01 16:02:35.741 +08:00] [INFO] [snap.rs:914] ["scan snapshot of one cf"] [size=10179124479] [key_count=521473] [cf=default] [snapshot=/var/lib/tikv/data/snap/gen_906_7_82250_(default|lock|>[2023/11/01 16:02:35.745 +08:00] [INFO] [snap.rs:914] ["scan snapshot of one cf"] [size=0] [key_count=0] [cf=lock] [snapshot=/var/lib/tikv/data/snap/gen_906_7_82250_(default|lock|write).sst] [regio>[2023/11/01 16:02:36.480 +08:00] [INFO] [snap.rs:914] ["scan snapshot of one cf"] [size=58926449] [key_count=1042946] [cf=write] [snapshot=/var/lib/tikv/data/snap/gen_906_7_82250_(default|lock|writ>[2023/11/01 16:02:36.481 +08:00] [INFO] [snap.rs:1088] ["scan snapshot"] [takes=132.578462083s] [size=7678717747] [key_count=1564419] [snapshot=/var/lib/tikv/data/snap/gen_906_7_82250_(default|lock>[2023/11/01 16:02:36.481 +08:00] [WARN] [peer_storage.rs:515] ["failed to try generating snapshot"] [request_peer=1537] [times=1] [peer_id=909] [region_id=906] [thread_id=0x5]
[2023/11/01 16:02:36.481 +08:00] [INFO] [peer_storage.rs:547] ["requesting snapshot"] [request_peer=1537] [request_index=0] [peer_id=909] [region_id=906] [thread_id=0x5]
[2023/11/01 16:03:15.884 +08:00] [INFO] [peer.rs:1307] ["deleting compacted snap file"] [snap_file=906_7_82250] [peer_id=909] [region_id=906] [thread_id=0x5]

[2023/11/01 16:03:28.468 +08:00] [INFO] [apply.rs:1691] ["execute admin command"] [command="cmd_type: ChangePeerV2 change_peer_v2 { changes { change_type: RemoveNode peer { id: 1537 store_id: 304 role: Learner } } }"] [index=95876] [term=7] [peer_id=909] [region_id=906] [thread_id=0x5]
[2023/11/01 16:03:28.468 +08:00] [INFO] [apply.rs:2285] ["exec ConfChangeV2"] [epoch="conf_ver: 24 version: 77"] [kind=Simple] [peer_id=909] [region_id=906] [thread_id=0x5]
[2023/11/01 16:03:28.468 +08:00] [INFO] [apply.rs:2466] ["conf change successfully"] ["current region"="id: 906 start_key: 7480000000000000FF7700000000000000F8 end_key: 7480000000000000FF7800000000000000F8 region_epoch { conf_ver: 25 version: 77 } peers { id: 909 store_id: 313 } peers { id: 967 store_id: 13 } peers { id: 994 store_id: 358 }"] ["original region"="id: 906 start_key: 7480000000000000FF7700000000000000F8 end_key: 7480000000000000FF7800000000000000F8 region_epoch { conf_ver: 24 version: 77 } peers { id: 909 store_id: 313 } peers { id: 967 store_id: 13 } peers { id: 994 store_id: 358 } peers { id: 1537 store_id: 304 role: Learner }"] [changes="[change_type: RemoveNode peer { id: 1537 store_id: 304 role: Learner }]"] [peer_id=909] [region_id=906] [thread_id=0x5]
[2023/11/01 16:03:28.469 +08:00] [INFO] [raft.rs:2660] ["switched to configuration"] [config="Configuration { voters: Configuration { incoming: Configuration { voters: {909, 994, 967} }, outgoing: Configuration { voters: {} } }, learners: {}, learners_next: {}, auto_leave: false }"] [raft_id=909] [region_id=906] [thread_id=0x5]
-->

### What version of TiKV are you using?
<!-- You can run `tikv-server --version` -->

v7.5.0

### Steps to reproduce
<!-- If possible, provide a recipe for reproducing the error. A complete runnable program is good. -->

It's an egde case, cannot reproduce reliably. Generally, run a workload with multiple hotspots.

### What did you expect?

No stale peer.

### What did happened?

Stale peer.

### PR DESCRIPTION
**Pull request by overvenus**
### What is changed and how it works?

Issue Number: Close #15919

What's Changed:

```commit-message
A "stale peer" refers to a peer that still exists on a TiKV node but
has been removed from the raft group, typically through a confchange
operation. TiKV performs regular checks and validations on its peers
to ensure that no such stale peer exists.

However, the current stale peer check is not enough when dealing with
uninitialized stale peers that its region has been merged. These
uninitialized stale peers are left indefinitely, consuming CPU, memory,
and blocking resolved ts.

This commit introduces an in-memory state for peers whose raft messages
create uninitialized stale peers. The stale peer then sends a
MsgCheckStalePeer message to the corresponding peer, validating whether
it should be removed.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

```release-note
Fix an issue that stale peers are left forever after region merge and blocks resolved ts
```

### CODE REVIEW

File: components/raftstore/src/store/worker/pd.rs [hunk 1]
```diff
@@ -1567,8 +1567,14 @@ where
...
                 Ok(None) => {
-                    // splitted Region has not yet reported to PD.
-                    // TODO: handle merge
+                    // Splitted region has not yet reported to PD.
+                    //
+                    // Or region has been merged. This case is handled by
+                    // message `MsgCheckStalePeer`, stale peers will be
+                    // removed eventually.
+                    PD_VALIDATE_PEER_COUNTER_VEC
+                        .with_label_values(&["region not found"])
```
Comment (Connor1996) [inline components/raftstore/src/store/worker/pd.rs:1576]: use static metrics？

**Review by Connor1996 — APPROVED** [review]
rest LGTM

**Review by SpadeA-Tang — APPROVED** [review]

File: components/raftstore/src/store/worker/pd.rs (hunk 1 above)
Comment (overvenus) [reply components/raftstore/src/store/worker/pd.rs:1576]: It's not on hot path, keep it consistent with others.

**Pull request merged**